### PR TITLE
vendor: update google/cadvisor and opencontainers/runc

### PR DESCRIPTION
**Pull request by giuseppe**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

vendor a new version of github.com/opencontainers/runc and adapt the callers to the new API to inherit fixes for cgroup v2

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20191118-cgroups-v2.md
```

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -76,16 +75,15 @@ require (
 	github.com/onsi/ginkgo v1.11.0
 	github.com/onsi/gomega v1.7.0
 	github.com/opencontainers/go-digest v1.0.0-rc1
-	github.com/opencontainers/runc v1.0.0-rc10
-	github.com/opencontainers/selinux v1.3.3
+	github.com/opencontainers/runc v1.0.0-rc9.0.20200519182809-b207d578ec2d
```
Comment (liggitt) [inline go.mod:78]: is this switching to an untagged release of runc?

Comment (dims) [reply go.mod:78]: @liggitt yep, we talked about it here - https://github.com/google/cadvisor/pull/2469#issuecomment-631435135 unfortunately runc rc11 is a while away

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 2]
```diff
@@ -500,8 +504,30 @@ func setResourcesV2(cgroupConfig *libcontainerconfigs.Cgroup) error {
...
+	cgroupConfig.Resources.Memory = -1
+	cgroupConfig.Resources.CpuWeight = 100
+	cgroupConfig.Resources.PidsLimit = -1
+
+	cpuset, err := ioutil.ReadFile(filepath.Join(cmutil.CgroupRoot, "cpuset.cpus.effective"))
+	if err != nil {
+		return fmt.Errorf("failed to read controllers from %q : %v", cmutil.CgroupRoot, err)
+	}
+	cgroupConfig.Resources.CpusetCpus = string(cpuset)
+	cgroupConfig.Resources.HugetlbLimit = []*libcontainerconfigs.HugepageLimit{
```
Comment (AkihiroSuda) [inline pkg/kubelet/cm/cgroup_manager_linux.go:525]: Is this something new in this PR?

**Comment by AkihiroSuda** [comment]
cc @kolyshkin

File: pkg/kubelet/cm/cgroup_manager_linux.go (hunk 2 above)
Comment (giuseppe) [reply pkg/kubelet/cm/cgroup_manager_linux.go]: Thanks for noticing it.  I've needed it for an older version of runc/libcontainer, it seems it is not needed anymore with master.  I'll drop this part.

File: go.mod (hunk 1 above)
Comment (liggitt) [reply go.mod:78]: it looks like this picks up some changes (https://github.com/opencontainers/runc/commit/4b4bc995ad4ee4679ffbe38f1672be01c24256fc) that were later reverted (https://github.com/opencontainers/runc/commit/a67dab0ac2dd3829858e35ba643613a1e358ba87)... it's unclear to me that this mid-tag version is suitable for us to pick up

Comment (giuseppe) [reply go.mod:78]: I've updated to a newer commit.  Fixes for cgroup v2 are still going to happen in libcontainer, it will likely take some more iterations to get it right.

Comment (liggitt) [reply go.mod:78]: > it will likely take some more iterations to get it right

Can you expand on this comment?

Is Kubernetes using cgroups v2 now? If so, it sounds like we're picking up changes we know are still in-progress, which doesn't seem desireable. If not, is it important for us to get this update now?

**Comment by dashpole** [comment]
updating cAdvisor to a non-tagged version should be fine.  We tag and update again before releases.

**Comment by liggitt** [comment]
I know we do that with cadvisor, I wasn't as sure we did that with runc. Are we using cgroups v2 currently?

**Comment by dashpole** [comment]
That is what the linked KEP is about.  I believe it is still in the process of going to alpha

**Comment by giuseppe** [comment]
> I know we do that with cadvisor, I wasn't as sure we did that with runc. Are we using cgroups v2 currently?

cgroup v2 support is targetting 1.19.

Unfortunately runc is still a moving target, there is still a lot of refactoring going on.  I've opened this PR to have some time to start using the current version and avoid introducing more changes later.

With this PR applied e2e node tests pass on a cgroup v2 host.

File: go.mod (hunk 1 above)
Comment (giuseppe) [reply go.mod]: sure, there is still a lot of work going on in the runc repository for supporting cgroup v2, the refactor affects also cgroup v1.  I've opened this PR as a checkpoint to avoid having to introduce more changes later when a release is finally cut (hopefully in time for 1.19) and have some time to use it

Comment (mrunalp) [reply go.mod]: We have a chicken and egg situation here. We want to validate the work done in runc for cgroups v2 so far with k8s, so we can go back and address any gaps or issues.

Comment (liggitt) [reply go.mod]: > the refactor affects also cgroup v1

That is my main concern. It's unclear to me what state this PR leaves the kubernetes master branch in, since the v2 changes we want to validate are entangled with changes to v1 code we currently depend on. I'm not in favor of experimental bumps of master to unstable versions of dependencies.

Comment (BenTheElder) [reply go.mod]: this is not really  a chicken and egg, if the work is experimental it can be tested with a patchset that is not in master.

**Comment by derekwaynecarr** [comment]
i am late to the discussion, but has runc situation changed?

**Comment by BenTheElder** [comment]
we were experiencing issues with container deletion leading to CI flakiness with runc at head both in KIND and the GCE tests FYI, https://github.com/opencontainers/runc/issues/2461

**Comment by mrunalp** [comment]
@derekwaynecarr We have an issue open and a plan to establish release cadence here https://github.com/opencontainers/runc/issues/2435. The next tag will `-rc91`.

I took a brief look at the issue @BenTheElder pointed out and not sure if we want to discuss that here, but it isn't clear to me what version was actually updated to. @BenTheElder says runc head, while the issue talks about runc `-rc10` which was released in January.

**Comment by BenTheElder** [comment]
it was "rc10+dev", though some (fine) runs were actually rc10.
there was some confusion around this, a bug in grabbing the version from
containerd vendor.conf in a nightly build pipeline led to us not using the
version they have pinned ...
@dims found the problem

**Comment by dims** [comment]
@giuseppe i pinged @jingyih in https://github.com/kubernetes/kubernetes/issues/91266 hopefully it will happen quick.

**Comment by giuseppe** [comment]
> @giuseppe i pinged @jingyih in #91266 hopefully it will happen quick.

sorry, I was wrong.  Latest released version v3.4.9 seems to work fine

**Comment by BenTheElder** [comment]
one odd failure I don't recognize, otherwise looks like flakes
>             Message: "an error on the server (\"Internal Error: failed to get imageFs stats: failed to get imageFs info: no imagefs label for configured runtime\") has prevented the request from succeeding (get nodes e2e-d0b27975fb-a7d53-minion-group-bnwj:10250)",
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/91366/pull-kubernetes-e2e-gce-ubuntu-containerd/1272538160076689408/
/retest

**Comment by BenTheElder** [comment]
/hold
The above mentioned failure is happening exclusively on this PR
https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&text=failed%20to%20get%20imageFs%20stats

**Comment by giuseppe** [comment]
> /hold
> The above mentioned failure is happening exclusively on this PR
> https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&text=failed%20to%20get%20imageFs%20stats

Thanks for letting me know.  I've fixed the issue, it was caused by a broken version of the patch "cadvisor: add Disk metrics with LocalStorageCapacityIsolation"

**Comment by BenTheElder** [comment]
/hold cancel
thanks!

**Comment by hakman** [comment]
@giuseppe sorry to drop in here out of the blue, but would be interested to the this fix from cAdvisor. Would it be possible to pull it in here, or I should wait for this to merge and come back with a new PR? Thanks! 😄 
https://github.com/google/cadvisor/pull/2586

**Comment by BenTheElder** [comment]
Passing tests now :upside_down_face:, thanks for fixing :+1: 
Poked #sig-node about this, particularly https://github.com/kubernetes/kubernetes/pull/91366#discussion_r432651487

**Comment by giuseppe** [comment]
> @giuseppe sorry to drop in here out of the blue, but would be interested to the this fix from cAdvisor. Would it be possible to pull it in here, or I should wait for this to merge and come back with a new PR? Thanks! smile
> [google/cadvisor#2586](https://github.com/google/cadvisor/pull/2586)

is it fine if I open a new PR to address that?  If this PR is not approved by tomorrow, then I guess I can reuse it and start again the CI dance :-)

**Comment by BenTheElder** [comment]
FWIW CI will run again on this PR before merge most likely anyhow (Kubernetes CI verifies HEAD of target branch merged with latest PR changes before actual merge, always), so I'd consider going ahead.

I only commented re: "CI is passing again" to note that the fix for the previous issue impacting only this PR seemed to be taking :-)

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 3]
```diff
@@ -514,26 +499,31 @@ func setResourcesV2(cgroupConfig *libcontainerconfigs.Cgroup) error {
...
+		},
+	}
 	if resourceConfig == nil {
 		return resources
 	}
 	if resourceConfig.Memory != nil {
 		resources.Memory = *resourceConfig.Memory
 	}
-	if libcontainercgroups.IsCgroup2UnifiedMode() {
-		resources.CpuWeight = getCpuWeight(resourceConfig.CpuShares)
```
Comment (derekwaynecarr) [inline pkg/kubelet/cm/cgroup_manager_linux.go:525]: why is this removed?  can you provide some context?

File: pkg/kubelet/cadvisor/cadvisor_linux.go [hunk 4]
```diff
@@ -96,7 +98,7 @@ func New(imageFsInfoProvider ImageFsInfoProvider, rootPath string, cgroupRoots [
 		cadvisormetrics.AppMetrics:              struct{}{},
 		cadvisormetrics.ProcessMetrics:          struct{}{},
 	}
-	if usingLegacyStats {
+	if usingLegacyStats || utilfeature.DefaultFeatureGate.Enabled(kubefeatures.LocalStorageCapacityIsolation) {
```
Comment (derekwaynecarr) [inline pkg/kubelet/cadvisor/cadvisor_linux.go:101]: this isnt clear to me.

**Review by derekwaynecarr — COMMENTED** [review]
will follow-up on slack, but i dont understand why max/weight were removed.

**Comment by hakman** [comment]
> is it fine if I open a new PR to address that? If this PR is not approved by tomorrow, then I guess I can reuse it and start again the CI dance :-)
That's fine by me. Any way it feels easier to you. Thanks again!

File: pkg/kubelet/cm/cgroup_manager_linux.go (hunk 3 above)
Comment (giuseppe) [reply pkg/kubelet/cm/cgroup_manager_linux.go:525]: I think I've done a mistake here.  libcontainer dropped CpuMax but CpuWeight is still needed and we need to do the conversion.  Thanks for noticing it, I'll fix it.

File: pkg/kubelet/cadvisor/cadvisor_linux.go (hunk 4 above)
Comment (giuseppe) [reply pkg/kubelet/cadvisor/cadvisor_linux.go:101]: this change is required for re-vendoring cAdvisor.  It won't collect any fs stats if there is no `cadvisormetrics.DiskUsageMetrics` specified.

This behaviour was changed with https://github.com/google/cadvisor/commit/e3affc1b79faa2901d7210948e92a5416166eb28

**Comment by giuseppe** [comment]
> That's fine by me. Any way it feels easier to you. Thanks again!

I've done a mistake in the PR so I will need anyway to iterate another time here.  I will take care of the re-vendoring as well :)

/hold

File: LICENSES/OWNERS [hunk 5]
```diff
@@ -1,4 +0,0 @@
-# See the OWNERS docs at https://go.k8s.io/owners
```
Comment (AkihiroSuda) [inline LICENSES/OWNERS:1]: Was this file removed by accident?

File: pkg/kubelet/dockershim/cm/BUILD [hunk 6]
```diff
@@ -41,6 +42,7 @@ go_library(
             "//pkg/kubelet/dockershim/libdocker:go_default_library",
             "//staging/src/k8s.io/apimachinery/pkg/util/version:go_default_library",
             "//staging/src/k8s.io/apimachinery/pkg/util/wait:go_default_library",
+            "//vendor/github.com/opencontainers/runc/libcontainer/cgroups:go_default_library",
```
Comment (AkihiroSuda) [inline pkg/kubelet/dockershim/cm/BUILD:45]: nit: indent

File: pkg/kubelet/cm/container_manager_linux_test.go [hunk 7]
```diff
@@ -89,6 +93,19 @@ func TestCgroupMountValidationMemoryMissing(t *testing.T) {
 	assert.Error(t, err)
 }
 
+func TestGetCpuWeight(t *testing.T) {
```
Comment (AkihiroSuda) [inline pkg/kubelet/cm/container_manager_linux_test.go:96]: moved by accident?

File: pkg/kubelet/dockershim/cm/BUILD (hunk 6 above)
Comment (giuseppe) [reply pkg/kubelet/dockershim/cm/BUILD:45]: it was generated by `hack/update-bazel.sh`

File: LICENSES/OWNERS (hunk 5 above)
Comment (giuseppe) [reply LICENSES/OWNERS:1]: yes thanks, restoring it

**Comment by dims** [comment]
one more thing to do ...
```
Use the given commands to remove pinned module versions that aren't actually used:
GO111MODULE=on go mod edit -dropreplace github.com/checkpoint-restore/go-criu
All pinned versions of checked dependencies match their preferred version.
```

**Comment by giuseppe** [comment]
> one more thing to do ...
> 
> ```
> Use the given commands to remove pinned module versions that aren't actually used:
> GO111MODULE=on go mod edit -dropreplace github.com/checkpoint-restore/go-criu
> All pinned versions of checked dependencies match their preferred version.
> ```

thanks!  Pushed a new version now

**Comment by giuseppe** [comment]
@liggitt @smarterclayton PTAL

**Comment by BenTheElder** [comment]
There was an experiment accidentally enabled on the IPv6 job. that job passing does not however block merges, and that's been fixed.
/retest

File: pkg/kubelet/cadvisor/cadvisor_linux.go (hunk 4 above)
Comment (liggitt) [reply pkg/kubelet/cadvisor/cadvisor_linux.go:101]: this feature is on by default... this effectively makes the default cadvisor config always include disk usage metrics. is that expected?

Comment (giuseppe) [reply pkg/kubelet/cadvisor/cadvisor_linux.go:101]: without this change, cAdvisor won't collect any fs stats: https://github.com/google/cadvisor/commit/e3affc1b79faa2901d7210948e92a5416166eb28#diff-061d1d5fc291bd4b18893f2ea01457ccR191-R194

That is expected given the change of behaviour in cAdvisor but it would be probably better to fix it in cAdvisor and have a way to retrieve only the global fs info.

**Comment by dims** [comment]
@giuseppe don't worry about the CI jobs that are not marked `Required` 

/approve
/lgtm

**Comment by dims** [comment]
@liggitt please peek when you get a chance.

File: hack/update-vendor-licenses.sh [hunk 8]
```diff
@@ -196,6 +196,10 @@ for PACKAGE in $(go list -m -json all | jq -r .Path | sort -f); do
     echo "${PACKAGE} doesn't exist in ${DEPS_DIR}, skipping" >&2
     continue
   fi
+  if [[ -z "$(find "${DEPS_DIR}/${PACKAGE}" -mindepth 1 -maxdepth 1 \! -name 'v[0-9]*')" ]]; then
```
Comment (liggitt) [inline hack/update-vendor-licenses.sh:199]: This can match false-positive dirs and skip requiring licenses for code we build/ship. For example, the https://github.com/kubernetes/api/tree/master/core package has a directory named v1, but it is not a versioned submodule.

We should only exclude directories which 1) contain no files and 2) only have subfolders which directly contain a go.mod file (the v[0-9] pattern doesn't matter)

**Comment by liggitt** [comment]
in general, my concerns in https://github.com/kubernetes/kubernetes/pull/91366#discussion_r432651487 still stand:

>> the refactor affects also cgroup v1
>
> That is my main concern. It's unclear to me what state this PR leaves the kubernetes master branch in, since the v2 changes we want to validate are entangled with changes to v1 code we currently depend on. I'm not in favor of experimental bumps of master to unstable versions of dependencies.

However, the sig-node leads are responsible for this component, so if this has their approval, I won't block it.

The [license-skipping](https://github.com/kubernetes/kubernetes/pull/91366#discussion_r444910570) comment is the only blocking comment I have.

File: go.mod [hunk 9]
```diff
@@ -429,7 +430,7 @@ replace (
 	golang.org/x/net => golang.org/x/net v0.0.0-20200324143707-d3edc9973b7e
 	golang.org/x/oauth2 => golang.org/x/oauth2 v0.0.0-20191202225959-858c2ad4c8b6
 	golang.org/x/sync => golang.org/x/sync v0.0.0-20190911185100-cd5d95a43a6e
-	golang.org/x/sys => golang.org/x/sys v0.0.0-20200201011859-915c9c3d4ccf // pinned to release-branch.go1.14-std
+	golang.org/x/sys => golang.org/x/sys v0.0.0-20200622214017-ed371f2e16b4 // pinned to release-branch.go1.14-std
```
Comment (liggitt) [inline go.mod:433]: revert to `release-branch.go1.14-std` (915c9c3d4ccf) if possible

Comment (giuseppe) [reply go.mod:433]: I needed to upgrade golang.org/x/sys because github.com/cilium/ebpf needs the definition for `unix.BPF_F_RDONLY_PROG`

File: hack/update-vendor-licenses.sh (hunk 8 above)
Comment (giuseppe) [reply hack/update-vendor-licenses.sh]: thanks, that is a much better idea.  I've changed the script to implement it

**Comment by liggitt** [comment]
looks unrelated, but these are flakes I haven't seen before:

@neolit123, fyi:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/91366/pull-kubernetes-integration/1275831207967330305
> === FAIL: cmd/kubeadm/test/cmd  (0.00s)
> FAIL	k8s.io/kubernetes/cmd/kubeadm/test/cmd	120.481s

@robscott, fyi:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/91366/pull-kubernetes-e2e-kind/1275831207984107522
> [sig-network] EndpointSlice should create Endpoints and EndpointSlices for Pods matching a Service
> test/e2e/network/endpointslice.go:162
Jun 24 17:08:14.774: Timed out waiting for matching EndpointSlices to exist: timed out waiting for the condition
test/e2e/network/endpointslice.go:321

/retest

**Comment by neolit123** [comment]
> === FAIL: cmd/kubeadm/test/cmd (0.00s)
> FAIL k8s.io/kubernetes/cmd/kubeadm/test/cmd 120.481s

logged https://github.com/kubernetes/kubeadm/issues/2197

**Comment by neolit123** [comment]
> > === FAIL: cmd/kubeadm/test/cmd (0.00s)
> > FAIL k8s.io/kubernetes/cmd/kubeadm/test/cmd 120.481s
> 
> logged [kubernetes/kubeadm#2197](https://github.com/kubernetes/kubeadm/issues/2197)

discussion is happening here:
https://github.com/kubernetes/kubernetes/issues/92360#issuecomment-649044024

**Comment by giuseppe** [comment]
tests are again green

**Comment by liggitt** [comment]
/approve
for vendor mechanics

/lgtm
based on sig-node lgtm+approve at https://github.com/kubernetes/kubernetes/pull/91366#issuecomment-645462686

File: vendor/github.com/moby/sys/mountinfo/LICENSE [hunk 10]
```diff
@@ -1,5 +1,3 @@
-= vendor/go.opencensus.io licensed under: =
```
Comment (liggitt) [inline vendor/github.com/moby/sys/mountinfo/LICENSE:1]: git marked this as a move, which masked the fact that something in this PR resulted in this license file getting dropped from LICENSES.
